File: java/src/test/java/quadTree/QuadTreeTest.java
package quadTree;

import org.junit.Test;

import java.io.FileInputStream;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import static org.junit.Assert.assertTrue;

public class QuadTreeTest {

	private static final String DATASET = "../data/dataset.txt";
	private static final String EXPECTED_DATASET = "../data/expected_dataset.txt";

	private Comparator<? super DataPoint> resultComparator = (o1, o2) -> {
        float r1 = (float)Math.sqrt(o1.x * o1.x + o1.y * o1.y);
        float r2 = (float)Math.sqrt(o2.x * o2.x + o2.y * o2.y);

        return Float.compare(r1, r2);
    };

	@Test
	public void testIntersect() throws Exception {
		List<DataPoint> dataset = DataSet.readDataPoints(new FileInputStream(DATASET));
		Rectangle rect = new Rectangle(-0.1f, -0.1f, 0.1f, 0.1f);

		List<DataPoint> expectedResult = DataSet.readDataPoints(new FileInputStream(EXPECTED_DATASET));
		List<DataPoint> actualResult = getIntersectedSetUsingQuadTree(dataset, rect);

		Collections.sort(actualResult, resultComparator);

		assertTrue(identical(actualResult, expectedResult));
	}

    /**
     * TODO implement me
     */
	private List<DataPoint> getIntersectedSetUsingQuadTree(List<DataPoint> dataset, Rectangle rect) {
		throw new UnsupportedOperationException("Not Yet Implemented");
	}

	private boolean identical(List<DataPoint> testSet1, List<DataPoint> testSet2) {
		if(testSet1.size() != testSet2.size()) {
            return false;
        }

		for (int i = 0; i < testSet1.size(); i++) {
			if (!testSet1.get(i).equals(testSet2.get(i))) {
				return false;
			}
		}

		return true;
	}
}

File: node/test/quadtreetest.js
"use strict";

var assert = require('assert');
var dataset = require('../lib/dataset');
var QuadTree = require('../lib/quadtree');
var Rectangle = require('../lib/datapoint');

describe('QuadTree', function () {
    it('Correctly Calculate the Intersection with a Rectangle', function () {
        let dataSet = dataset("../data/dataset.txt");
        let rect = new Rectangle(-0.1, -0.1, 0.1, 0.1);

        let expectedResult = dataset("../data/expected_dataset.txt");
        let actualResult = getIntersectedSetUsingQuadTree(dataSet, rect);

        actualResult.sort(compareDataPoints)
        expectedResult.sort(compareDataPoints)
        identical(actualResult, expectedResult);
    });

    function getIntersectedSetUsingQuadTree(dataSet, rect) {
        throw new Error("Not Yet Implemented");
    }

    function identical(actual, expected) {
        assert.equal(actual.length, expected.length, "Different Number of DataPoints in result arrays");

        for (let i = 0; i < actual.length; i++) {
            let currentActual = actual[i];
            let currentExpected = expected[i];
            assert.equal(currentActual.x, currentExpected.x, `Datapoints at position ${i} have different x values [${currentActual.x}] and [${currentExpected.x}]`);
            assert.equal(currentActual.y, currentExpected.y, `Datapoints at position ${i} have different x values [${currentActual.y}] and [${currentExpected.y}]`);
        }
    }

    function compareDataPoints(o1, o2) {
        let r1 = Math.sqrt(o1.x * o1.x + o1.y * o1.y);
        let r2 = Math.sqrt(o2.x * o2.x + o2.y * o2.y);

        // TODO this probably doesn't handle NaN vs -0.0 vs 0.0 correctly
        if (r1 < r2) {
            return -1
        } else if (r1 > r2) {
            return 1;
        } else {
            return 0;
        }
    }

});